src/main/java/seedu/address/model/person/PersonMatchesKeywordsPredicate.java

```
@@ -0,0 +1,361 @@
+        }
+
+        return predicates;
+    }
+
+    /**
+     * Returns a composed predicate that represents matching all of the given predicates.
+     *
+     * @param predicates A list of predicates to match.
+     * @return A predicate that returns true only if all predicates test true on a person.
+     */
+    private Predicate<Person> matchAll(List<Predicate<Person>> predicates) {
+        return predicates.stream().reduce(x -> true, Predicate::and);
+    }
+
+    /**
+     * Returns a composed predicate that represents matching any one of the given predicates.
+     *
+     * @param predicates A list of predicates to match.
+     * @return A predicate that returns true if any one of the predicates test true on a person.
+     */
+    private Predicate<Person> matchAny(List<Predicate<Person>> predicates) {
+        return predicates.stream().reduce(x -> false, Predicate::or);
+    }
+
+    /**
+     * Returns a composed predicate that represents matching none of the given predicates.
+     *
+     * @param predicates A list of predicates to match.
+     * @return A predicate that returns true only if none of the predicates test true on a person.
+     */
+    private Predicate<Person> matchNone(List<Predicate<Person>> predicates) {
+        return matchAny(predicates).negate();
+    }
+
+    /**
+     * Evaluates this predicate on the given person.
+     *
+     * @param person The person to test on.
+     * @return True if the person matches the predicate, otherwise false
+     */
+    @Override
+    public boolean test(Person person) {
+        List<Predicate<Person>> predicates = getPredicates();
+        switch (condition) {
+        case ALL:
+            return matchAll(predicates).test(person);
+        case ANY:
+            return matchAny(predicates).test(person);
+        case NONE:
+            return matchNone(predicates).test(person);
+        default:
+            // should not reach here
+            return false;
```

maybe can throw an `exception` or use `assert` instead of `return false`?